File: samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/BlobItemViewHolder.java
package com.azure.android.storage.sample;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Environment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.content.FileProvider;
import androidx.lifecycle.LifecycleOwner;
import androidx.recyclerview.widget.RecyclerView;

import com.azure.android.storage.blob.StorageBlobClient;
import com.azure.android.storage.blob.models.BlobItem;
import com.azure.android.storage.sample.config.StorageConfiguration;

import java.io.File;

public class BlobItemViewHolder extends RecyclerView.ViewHolder {
    private static final String TAG = BlobItemViewHolder.class.getSimpleName();
    private final TextView blobName;

    private BlobItemViewHolder(@NonNull View itemView) {
        super(itemView);
        this.blobName = itemView.findViewById(R.id.blob_name);
    }

    public static BlobItemViewHolder create(ViewGroup parent, StorageBlobClient storageBlobClient) {
        StorageConfiguration storageConfiguration = StorageConfiguration.create(parent.getContext());
        View blobItemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.blob_item, parent, false);

        blobItemView.setOnClickListener(view -> {
            String blobName = (String) ((TextView) view.findViewById(R.id.blob_name)).getText();
            String containerName = storageConfiguration.getContainerName();

            Toast.makeText(parent.getContext(), "Downloading " + blobName, Toast.LENGTH_SHORT).show();

            View mainActivityView = (View) parent.getParent().getParent().getParent();
            ProgressBar progressBar = mainActivityView.findViewById(R.id.download_progress_bar);

            showProgress(mainActivityView);

            File path = Environment.getExternalStorageDirectory();
            File file = new File(path, blobName);

            try {
                storageBlobClient.download(parent.getContext(), containerName, blobName, file)
                    .enqueue()
                    .observe((LifecycleOwner) parent.getContext(), new TransferObserver() {
                        @Override
                        public void onStart(String transferId) {
                            Log.i(TAG, "onStart() for transfer with ID: " + transferId);

                            Button cancelButton = mainActivityView.findViewById(R.id.cancel_button);

                            cancelButton.setOnClickListener(v -> {
                                storageBlobClient.cancel(parent.getContext(), transferId);
                                hideProgress(mainActivityView);
                                Toast.makeText(parent.getContext(), "Download cancelled", Toast.LENGTH_SHORT)
                                    .show();
                            });
                        }

                        @Override
                        public void onProgress(String transferId, long totalBytes, long bytesTransferred) {
                            Log.i(TAG, "onProgress(" + totalBytes + ", " + bytesTransferred +
                                ") for transfer with ID:" + transferId);


                            if (progressBar != null) {
                                if (progressBar.getProgress() == 0) {
                                    progressBar.setMax((int) totalBytes);
                                }

                                progressBar.setProgress((int) bytesTransferred);
                            }
                        }

                        @Override
                        public void onSystemPaused(String transferId) {
                            Log.i(TAG, "onSystemPaused() for transfer with ID: " + transferId);
                        }

                        @Override
                        public void onResume(String transferId) {
                            Log.i(TAG, "onResumed() for transfer with ID: " + transferId);
                        }

                        @Override
                        public void onComplete(String transferId) {
                            Log.i(TAG, "onCompleted() for transfer with ID: " + transferId);

                            if (progressBar != null) {
                                progressBar.setProgress(progressBar.getMax());
                            }

                            hideProgress(mainActivityView);
                            Toast.makeText(parent.getContext(), "Download complete", Toast.LENGTH_SHORT).show();
                            showFileIntent(parent.getContext(), file);

                            if (progressBar != null) {
                                progressBar.setProgress(0);
                            }
                        }

                        @Override
                        public void onError(String transferId, String errorMessage) {
                            Log.i(TAG, "onError() for transfer with ID: " + transferId + " -> : " + errorMessage);

                            hideProgress(mainActivityView);
                            Toast.makeText(parent.getContext(), "Download failed", Toast.LENGTH_SHORT).show();

                            if (progressBar != null) {
                                progressBar.setProgress(0);
                            }
                        }
                    });
            } catch (Exception ex) {
                Log.e(TAG, "Blob download failed: ", ex);
            }
        });

        return new BlobItemViewHolder(blobItemView);
    }

    public void bind(BlobItem blobItem) {
        if (blobItem != null) {
            this.blobName.setText(blobItem.getName() != null ? blobItem.getName() : "loading...");
        }
    }

    private static void showProgress(View mainActivityView) {
        mainActivityView.findViewById(R.id.download_background).setVisibility(View.VISIBLE);
        mainActivityView.findViewById(R.id.download_progress_bar).setVisibility(View.VISIBLE);
        mainActivityView.findViewById(R.id.download_buttons).setVisibility(View.VISIBLE);
        mainActivityView.findViewById(R.id.cancel_button).setVisibility(View.VISIBLE);
    }

    private static void hideProgress(View mainActivityView) {
        mainActivityView.findViewById(R.id.cancel_button).setVisibility(View.GONE);
        mainActivityView.findViewById(R.id.download_buttons).setVisibility(View.GONE);
        mainActivityView.findViewById(R.id.download_progress_bar).setVisibility(View.GONE);
        mainActivityView.findViewById(R.id.download_background).setVisibility(View.GONE);
    }

    private static void showFileIntent(Context context, File file) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        Uri fileUri =
            FileProvider.getUriForFile(context,context.getApplicationContext().getPackageName() + ".provider", file);
        String contentType = context.getContentResolver().getType(fileUri);
        intent.setDataAndType(fileUri, contentType);
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        context.startActivity(intent);
    }
}

File: samples/sample-app-storage-kotlin/src/main/java/com/azure/android/storage/sample/kotlin/BlobItemViewHolder.kt
package com.azure.android.storage.sample.kotlin

import android.content.Context
import android.content.Intent
import android.os.Environment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.core.content.FileProvider
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.RecyclerView
import com.azure.android.storage.blob.StorageBlobAsyncClient
import com.azure.android.storage.blob.models.BlobItem
import com.azure.android.storage.sample.kotlin.config.StorageConfiguration
import java.io.File

class BlobItemViewHolder private constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
    private val blobName: TextView = itemView.findViewById(R.id.blob_name)

    fun bind(blobItem: BlobItem?) {
        if (blobItem != null) {
            blobName.text = if (blobItem.name != null) blobItem.name else "loading..."
        }
    }

    companion object {
        private val TAG = BlobItemViewHolder::class.java.simpleName

        fun create(parent: ViewGroup, storageBlobAsyncClient: StorageBlobAsyncClient): BlobItemViewHolder {
            val storageConfiguration: StorageConfiguration = StorageConfiguration.create(parent.context)
            val blobItemView = LayoutInflater.from(parent.context).inflate(R.layout.blob_item, parent, false)
            blobItemView.setOnClickListener { view: View ->
                val blobName = (view.findViewById<View>(R.id.blob_name) as TextView).text as String
                Toast.makeText(parent.context, "Downloading $blobName", Toast.LENGTH_SHORT).show()

                val mainActivityView = parent.parent.parent.parent as View
                val progressBar = mainActivityView.findViewById<ProgressBar>(R.id.download_progress_bar)
                showProgress(mainActivityView)

                val path = Environment.getExternalStorageDirectory()
                val file = File(path, blobName)
                try {
                    val containerName: String? = storageConfiguration.containerName
                    storageBlobAsyncClient.download(parent.context, containerName, blobName, file)
                        .observe((parent.context as LifecycleOwner), object : TransferObserver {
                            override fun onStart(transferId: Long) {
                                Log.i(TAG, "onStart() for transfer with ID: $transferId")
                                val cancelButton = mainActivityView.findViewById<Button>(R.id.cancel_button)
                                cancelButton.setOnClickListener {
                                    storageBlobAsyncClient.cancel(parent.context, transferId)
                                    hideProgress(mainActivityView)
                                    Toast.makeText(parent.context, "Download cancelled", Toast.LENGTH_SHORT)
                                        .show()
                                }
                            }

                            override fun onProgress(transferId: Long, totalBytes: Long, bytesTransferred: Long) {
                                Log.i(TAG, "onProgress(" + totalBytes + ", " + bytesTransferred +
                                    ") for transfer with ID:" + transferId)
                                if (progressBar != null) {
                                    if (progressBar.progress == 0) {
                                        progressBar.max = totalBytes.toInt()
                                    }
                                    progressBar.progress = bytesTransferred.toInt()
                                }
                            }

                            override fun onSystemPaused(transferId: Long) {
                                Log.i(TAG, "onSystemPaused() for transfer with ID: $transferId")
                            }

                            override fun onResume(transferId: Long) {
                                Log.i(TAG, "onResumed() for transfer with ID: $transferId")
                            }

                            override fun onComplete(transferId: Long) {
                                Log.i(TAG, "onCompleted() for transfer with ID: $transferId")
                                if (progressBar != null) {
                                    progressBar.progress = progressBar.max
                                }
                                hideProgress(mainActivityView)
                                Toast.makeText(parent.context, "Download complete", Toast.LENGTH_SHORT).show()
                                showFileIntent(parent.context, file)
                                if (progressBar != null) {
                                    progressBar.progress = 0
                                }
                            }

                            override fun onError(transferId: Long, errorMessage: String?) {
                                Log.i(TAG, "onError() for transfer with ID: $transferId -> : $errorMessage")
                                hideProgress(mainActivityView)
                                Toast.makeText(parent.context, "Download failed", Toast.LENGTH_SHORT).show()
                                if (progressBar != null) {
                                    progressBar.progress = 0
                                }
                            }
                        })
                } catch (ex: Exception) {
                    Log.e(TAG, "Blob download failed: ", ex)
                }
            }
            return BlobItemViewHolder(blobItemView)
        }

        private fun showProgress(mainActivityView: View) {
            mainActivityView.findViewById<View>(R.id.download_background).visibility = View.VISIBLE
            mainActivityView.findViewById<View>(R.id.download_progress_bar).visibility = View.VISIBLE
            mainActivityView.findViewById<View>(R.id.download_buttons).visibility = View.VISIBLE
            mainActivityView.findViewById<View>(R.id.cancel_button).visibility = View.VISIBLE
        }

        private fun hideProgress(mainActivityView: View) {
            mainActivityView.findViewById<View>(R.id.cancel_button).visibility = View.GONE
            mainActivityView.findViewById<View>(R.id.download_buttons).visibility = View.GONE
            mainActivityView.findViewById<View>(R.id.download_progress_bar).visibility = View.GONE
            mainActivityView.findViewById<View>(R.id.download_background).visibility = View.GONE
        }

        private fun showFileIntent(context: Context, file: File) {
            val intent = Intent(Intent.ACTION_VIEW)
            val fileUri = FileProvider.getUriForFile(context, "com.azure.android.storage.sample.kotlin" + ".provider", file)
            val contentType = context.contentResolver.getType(fileUri)
            intent.setDataAndType(fileUri, contentType)
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            context.startActivity(intent)
        }
    }
}

